Microsoft.Azure.Cosmos/src/Handler/TelemetryHandler.cs

```
@@ -0,0 +1,86 @@
+﻿//------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+//------------------------------------------------------------
+
+namespace Microsoft.Azure.Cosmos.Handlers
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.IO;
+    using System.Net.Http;
+    using System.Runtime.InteropServices;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos.Core.Trace;
+    using Microsoft.Azure.Cosmos.CosmosElements;
+    using Microsoft.Azure.Cosmos.Handler;
+    using Microsoft.Azure.Cosmos.Tracing.TraceData;
+    using Newtonsoft.Json.Linq;
+
+    internal class TelemetryHandler : RequestHandler
+    {
+        private readonly ClientTelemetry telemetry;
+
+        public TelemetryHandler(ClientTelemetry telemetry)
+        {
+            this.telemetry = telemetry ?? throw new ArgumentNullException(nameof(telemetry));
+        }
+
+        public override async Task<ResponseMessage> SendAsync(
+            RequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            ResponseMessage response = await base.SendAsync(request, cancellationToken);
+            if (this.IsAllowed(request))
+            {
+                this.telemetry
+                .Collect(
+                        cosmosDiagnostics: response.Diagnostics,
+                        statusCode: response.StatusCode,
+                        responseSizeInBytes: this.GetPayloadSize(response),
+                        containerId: request.ContainerId,
+                        databaseId: request.DatabaseId,
+                        operationType: request.OperationType,
+                        resourceType: request.ResourceType,
+                        consistencyLevel: this.GetConsistencyLevel(request),
+                        requestCharge: response.Headers.RequestCharge);
```

What are the error/exception handling semantics? 

Won't an exception from Collect impact the mainline folow?

---

till now I saw few failure while recording values so handles the same but as I mentioned above if you can see any scope of exception. I can put one more try-catch above it just to absorb any kind of exception thrown by collect function

---

Observations from runs is a very limited sample and doesn't cover the full possible set. 

/cc: @j82w

---

I think we should throw an exception on collect path instead of eating the exception. Customers can disabled telemetry to bypass it. This way it is very clear than is an issue that needs to be fixed. Only have an exception show up in the logs is going to be difficult to discover and troubleshoot.

---

As per offline discussion with @kirankumarkolli , absorbing this exception as of now. Added this point in this workitem : https://github.com/Azure/azure-cosmos-dotnet-v3/issues/2604